api/lightning/node_cln.py

```diff
@@ -0,0 +1,408 @@
+        210: "Payment timed out without a payment in progress.",
+    }
+
+    @classmethod
+    def decode_payreq(cls, invoice):
+        """Decodes a lightning payment request (invoice)"""
+        request = noderpc.DecodeBolt11Request(bolt11=invoice)
+
+        response = cls.stub.DecodeBolt11(request)
+        return response
+
+    @classmethod
+    def estimate_fee(cls, amount_sats, target_conf=2, min_confs=1):
+        """Returns estimated fee for onchain payouts"""
+        # feerate estimaes work a bit differently in cln see https://lightning.readthedocs.io/lightning-feerates.7.html
+        request = noderpc.FeeratesRequest(style="PERKB")
+
+        response = cls.stub.Feerates(request)
+
+        # "opening" -> ~12 block target
+        return {
+            "mining_fee_sats": response.onchain_fee_estimates.opening_channel_satoshis,
+            "mining_fee_rate": response.perkb.opening/1000,
+        }
+
+    wallet_balance_cache = {}
+
+    @ring.dict(wallet_balance_cache, expire=10)  # keeps in cache for 10 seconds
+    @classmethod
+    def wallet_balance(cls):
+        """Returns onchain balance"""
+        request = noderpc.ListfundsRequest()
+
+        response = cls.stub.ListFunds(request)
+
+        unconfirmed_balance = 0
+        confirmed_balance = 0
+        total_balance = 0
+        for utxo in response.outputs:
+                if not utxo.reserved:
+                    if utxo.status == 0: # UNCONFIRMED
+                        unconfirmed_balance += utxo.amount_msat.msat // 1_000
+                        total_balance += utxo.amount_msat.msat // 1_000
+                    elif utxo.status == 1: # CONFIRMED
+                        confirmed_balance += utxo.amount_msat.msat // 1_000
+                        total_balance += utxo.amount_msat.msat // 1_000
+
+        return {
+            "total_balance": total_balance,
+            "confirmed_balance": confirmed_balance,
+            "unconfirmed_balance": unconfirmed_balance,
+        }
+
+    channel_balance_cache = {}
+
+    @ring.dict(channel_balance_cache, expire=10)  # keeps in cache for 10 seconds
+    @classmethod
+    def channel_balance(cls):
+        """Returns channels balance"""
+        request = noderpc.ListpeersRequest()
+
+        response = cls.stub.ListPeers(request)
+
+        local_balance_sat = 0
+        remote_balance_sat = 0
+        for peer in response.peers:
+            for channel in peer.channels:
+                if channel.state == 2: # CHANNELD_NORMAL
+                    local_balance_sat += channel.to_us_msat.msat // 1_000
+                    remote_balance_sat += (channel.total_msat.msat-channel.to_us_msat.msat) // 1_000
+        
+        # TODO no idea what is meant by unsettled/pending balance here exactly
```

Comment:
"Unsettled balance" is the total amount on incoming HTLCs on hold. "Pending balance" is the total "in_flight" amount (outgoing payments in transition).